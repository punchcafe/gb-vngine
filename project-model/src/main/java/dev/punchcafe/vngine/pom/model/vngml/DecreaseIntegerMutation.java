package dev.punchcafe.vngine.pom.model.vngml;

import dev.punchcafe.vngine.pom.model.vngpl.variable.integer.IntegerGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.integer.IntegerLiteral;
import lombok.*;

@Getter
@Builder
@EqualsAndHashCode
@ToString
public class DecreaseIntegerMutation implements GameStateMutationExpression {
    private IntegerGameVariable variableToModify;
    private IntegerLiteral decreaseBy;

    @Override
    public <T> T acceptVisitor(GameStateMutationExpressionVisitor<T> visitor) {
        return visitor.visitDecreaseIntegerMutation(this);
    }

    @Override
    public String asVngQL() {
        return "decrease " + variableToModify.asVngQL() + " by " + decreaseBy.asVngQL();
    }
}
