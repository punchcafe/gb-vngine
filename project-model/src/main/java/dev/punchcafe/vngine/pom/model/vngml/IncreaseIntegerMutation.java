package dev.punchcafe.vngine.pom.model.vngml;

import dev.punchcafe.vngine.pom.model.vngpl.variable.integer.IntegerGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.integer.IntegerLiteral;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@Builder
@EqualsAndHashCode
@ToString
public class IncreaseIntegerMutation implements GameStateMutationExpression {
    private IntegerGameVariable variableToModify;
    private IntegerLiteral increaseBy;

    @Override
    public <T> T acceptVisitor(GameStateMutationExpressionVisitor<T> visitor) {
        return visitor.visitIncreaseIntegerMutation(this);
    }

    @Override
    public String asVngQL() {
        return "increase " + variableToModify.asVngQL() + " by " + increaseBy.asVngQL();
    }
}
