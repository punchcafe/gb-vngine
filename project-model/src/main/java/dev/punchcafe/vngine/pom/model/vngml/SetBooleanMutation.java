package dev.punchcafe.vngine.pom.model.vngml;

import dev.punchcafe.vngine.pom.model.vngpl.variable.bool.BoolGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.bool.BooleanLiteral;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@Builder
@EqualsAndHashCode
@ToString
public class SetBooleanMutation implements GameStateMutationExpression {
    private BoolGameVariable variableToMutate;
    // TODO: consider allowing this to be other game variables
    private BooleanLiteral changeValue;

    @Override
    public <T> T acceptVisitor(GameStateMutationExpressionVisitor<T> visitor) {
        return visitor.visitSetBooleanMutation(this);
    }

    @Override
    public String asVngQL() {
        return "set " + variableToMutate.asVngQL() + " to " + changeValue.asVngQL();
    }
}
