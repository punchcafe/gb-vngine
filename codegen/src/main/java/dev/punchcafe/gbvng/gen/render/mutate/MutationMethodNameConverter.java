package dev.punchcafe.gbvng.gen.render.mutate;

import dev.punchcafe.gbvng.gen.adapter.CNameSanitiser;
import dev.punchcafe.vngine.pom.model.VariableTypes;
import dev.punchcafe.vngine.pom.model.vngml.*;
import dev.punchcafe.vngine.pom.model.vngpl.variable.GameVariableLevel;
import dev.punchcafe.vngine.pom.model.vngpl.variable.bool.BoolGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.integer.IntegerGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.string.StringGameVariable;

public class MutationMethodNameConverter implements GameStateMutationExpressionVisitor<String> {

    private static MutationMethodNameConverter SINGLETON = new MutationMethodNameConverter();

    private final String mutateStringTemplate = "set_%s_to_%s";
    private final String mutateBoolTemplate = "set_%s_to_%s";
    private final String increaseIntTemplate = "inc_%s_to_%d";
    private final String decreaseIntTemplate = "dec_%s_to_%d";

    public static String convert(final GameStateMutationExpression expression) {
        return expression.acceptVisitor(SINGLETON);
    }

    @Override
    public String visitSetBooleanMutation(SetBooleanMutation setBooleanMutation) {
        final var sanitisedVarName = sanitiseVariableName(setBooleanMutation.getVariableToMutate());
        final var booleanLiteral = setBooleanMutation.getChangeValue()
                .toString()
                .toLowerCase();

        return String.format(mutateBoolTemplate, sanitisedVarName, booleanLiteral);
    }

    @Override
    public String visitSetStringMutation(SetStringMutation setStringMutation) {
        throwIfChapterVariable(setStringMutation.getVariableToMutate().getGameVariableLevel());
        final var sanitisedVarName = sanitiseVariableName(setStringMutation.getVariableToMutate());

        final var sanitisedStringLiteral = CNameSanitiser.sanitiseStringLiteral(setStringMutation.getChangeValue().getValue());
        return String.format(mutateStringTemplate, sanitisedVarName, sanitisedStringLiteral);
    }

    @Override
    public String visitIncreaseIntegerMutation(IncreaseIntegerMutation increaseIntegerMutation) {
        throwIfChapterVariable(increaseIntegerMutation.getVariableToModify().getGameVariableLevel());

        final var sanitisedVarName = sanitiseVariableName(increaseIntegerMutation.getVariableToModify());
        final var increaseValue = increaseIntegerMutation.getIncreaseBy().getValue();

        return String.format(increaseIntTemplate, sanitisedVarName, increaseValue);
    }

    private String sanitiseVariableName(StringGameVariable variable) {
        return CNameSanitiser.sanitiseVariableName(variable.getVariableName(),
                VariableTypes.STR,
                variable.getGameVariableLevel());
    }

    private String sanitiseVariableName(BoolGameVariable variable) {
        return CNameSanitiser.sanitiseVariableName(variable.getVariableName(),
                VariableTypes.BOOL,
                variable.getGameVariableLevel());
    }

    private String sanitiseVariableName(IntegerGameVariable variable) {
        return CNameSanitiser.sanitiseVariableName(variable.getVariableName(),
                VariableTypes.INT,
                variable.getGameVariableLevel());
    }

    @Override
    public String visitDecreaseIntegerMutation(DecreaseIntegerMutation decreaseIntegerMutation) {
        final var sanitisedVarName = sanitiseVariableName(decreaseIntegerMutation.getVariableToModify());
        final var increaseValue = decreaseIntegerMutation.getDecreaseBy().getValue();

        return String.format(decreaseIntTemplate, sanitisedVarName, increaseValue);
    }

    private void throwIfChapterVariable(final GameVariableLevel level) {
        if (level != GameVariableLevel.GAME) {
            throw new UnsupportedOperationException();
        }
    }
}
