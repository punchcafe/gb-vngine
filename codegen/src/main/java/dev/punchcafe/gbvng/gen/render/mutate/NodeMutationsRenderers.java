package dev.punchcafe.gbvng.gen.render.mutate;

import dev.punchcafe.gbvng.gen.adapter.NodeMutationsArrayName;
import dev.punchcafe.gbvng.gen.render.ComponentRenderer;
import dev.punchcafe.vngine.pom.model.Node;
import dev.punchcafe.vngine.pom.model.ProjectObjectModel;
import dev.punchcafe.vngine.pom.parse.vngml.GameStateMutationExpressionParser;
import lombok.Builder;

import java.util.List;
import java.util.stream.Collectors;

import static dev.punchcafe.gbvng.gen.render.ComponentRendererNames.NODE_MUTATION_RENDERER_NAME;
import static dev.punchcafe.gbvng.gen.render.mutate.GameStateMutationRenderer.GAME_STATE_MUTATION_RENDERER;
import static java.util.stream.Collectors.joining;

@Builder
public class NodeMutationsRenderers implements ComponentRenderer {

    private static final GameStateMutationExpressionParser mutationParser = new GameStateMutationExpressionParser();

    private final ProjectObjectModel<?> gameConfig;

    @Override
    public String render() {
        return gameConfig.getChapterConfigs().get(0)
                .getNodes()
                .stream()
                .map(this::renderNodeMutations)
                .collect(Collectors.joining("\n"));
    }

    @Override
    public List<String> dependencies() {
        return List.of(GAME_STATE_MUTATION_RENDERER);
    }

    @Override
    public String componentName() {
        return NODE_MUTATION_RENDERER_NAME;
    }

    private String renderNodeMutations(final Node node) {

        if(node.getGameStateModifiers() == null || node.getGameStateModifiers().isEmpty()){
            return "";
        }

        final var mutationArrayPointerName = NodeMutationsArrayName.getMutationArrayForNode(node);
        final var methodReferencesArray = node.getGameStateModifiers().stream()
                .map(mutationParser::parseExpression)
                .map(MutationMethodNameConverter::convert)
                .map(methodName -> "&" + methodName)
                .collect(joining(",", "{", "}"));
        final var numberOfMutations = node.getGameStateModifiers().size();

        return String.format("void (*%s[%d])(struct GameState *) = %s;", mutationArrayPointerName,
                numberOfMutations,
                methodReferencesArray);
    }

}
