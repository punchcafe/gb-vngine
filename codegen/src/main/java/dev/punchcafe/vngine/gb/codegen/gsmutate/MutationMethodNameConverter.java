package dev.punchcafe.vngine.gb.codegen.gsmutate;

import dev.punchcafe.vngine.pom.model.vngml.*;
import dev.punchcafe.vngine.pom.model.vngpl.variable.GameVariableLevel;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static java.nio.charset.StandardCharsets.UTF_8;

public class MutationMethodNameConverter implements GameStateMutationExpressionVisitor<String> {

    private static Map<Character, String> ALIASES = new HashMap<>() {{
        this.put('-', "__DASH__");
        this.put('.', "__DOT__");
        this.put('~', "__TILDE__");
        this.put(',', "__COMMA__");
        this.put(':', "__COLON__");
        this.put('/', "__FSLASH__");
        this.put('?', "__QMARK__");
        this.put('#', "__HASHTAG__");
        this.put('[', "__LSQBRA__");
        this.put(']', "__RSBRA__");
        this.put('@', "__ATSYM__");
        this.put('!', "__EXCLMARK__");
        this.put('$', "__DOLLAR__");
        this.put('&', "__ANDSYM__");
        this.put('\'', "__SINGQUOT__");
        this.put('(', "__LBRA__");
        this.put(')', "__RBRA__");
        this.put('*', "__ASTERISK__");
        this.put('+', "__PLUS__");
        this.put(';', "__SEMICOL__");
        this.put('%', "__PERCENT__");
        this.put('=', "__EQUALS__");
    }};

    private static MutationMethodNameConverter SINGLETON = new MutationMethodNameConverter();

    private final String mutateStringTemplate = "set_glob_str_var_%s_to_%s";
    private final String mutateBoolTemplate = "set_glob_bool_var_%s_to_%s";
    private final String increaseIntTemplate = "inc_glob_int_var_%s_to_%d";
    private final String decreaseIntTemplate = "dec_glob_int_var_%s_to_%d";

    public static String convert(final GameStateMutationExpression expression) {
        return expression.acceptVisitor(SINGLETON);
    }

    private String encodeUrlString(final String str) {
        try {
            return URLEncoder.encode(str, UTF_8.toString());
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException();
        }
    }

    @Override
    public String visitSetBooleanMutation(SetBooleanMutation setBooleanMutation) {
        final var sanitisedVarName = sanitiseVariableName(setBooleanMutation
                .getVariableToMutate()
                .getVariableName());
        final var booleanLiteral = setBooleanMutation.getChangeValue()
                .toString()
                .toLowerCase();

        return String.format(mutateBoolTemplate, sanitisedVarName, booleanLiteral);
    }

    @Override
    public String visitSetStringMutation(SetStringMutation setStringMutation) {
        throwIfChapterVariable(setStringMutation.getVariableToMutate().getGameVariableLevel());
        final var sanitisedVarName = sanitiseVariableName(setStringMutation
                .getVariableToMutate()
                .getVariableName());

        final var sanitisedStringLiteral = sanitiseStringLiteral(setStringMutation.getChangeValue().getValue());
        return String.format(mutateStringTemplate, sanitisedVarName, sanitisedStringLiteral);
    }

    private String sanitiseStringLiteral(String value) {
        return Optional.of(value)
                .map(this::encodeUrlString)
                .map(this::replaceUrlCharacterDelimiters)
                .map(this::replaceIllegalCFunctionCharacters)
                .get();
    }

    private String replaceIllegalCFunctionCharacters(String s) {
        final var chars = s.toCharArray();
        final var stringBuilder = new StringBuilder();
        for (final var charr : chars) {
            Optional.ofNullable(ALIASES.get(charr))
                    .or(() -> Optional.of(String.valueOf(charr)))
                    .ifPresent(stringBuilder::append);
        }
        return stringBuilder.toString();
    }

    private String replaceUrlCharacterDelimiters(final String string) {
        return string.replaceAll("%(\\d{2})", "__URLCHAR_$1__");
    }

    @Override
    public String visitIncreaseIntegerMutation(IncreaseIntegerMutation increaseIntegerMutation) {
        throwIfChapterVariable(increaseIntegerMutation.getVariableToModify().getGameVariableLevel());

        final var sanitisedVarName = sanitiseVariableName(increaseIntegerMutation
                .getVariableToModify()
                .getVariableName());
        final var increaseValue = increaseIntegerMutation.getIncreaseBy().getValue();

        return String.format(increaseIntTemplate, sanitisedVarName, increaseValue);
    }

    private String sanitiseVariableName(String variableName) {
        return Optional.of(variableName)
                .map(str -> str.replace("-", "_"))
                .get();
    }

    @Override
    public String visitDecreaseIntegerMutation(DecreaseIntegerMutation decreaseIntegerMutation) {
        final var sanitisedVarName = sanitiseVariableName(decreaseIntegerMutation
                .getVariableToModify()
                .getVariableName());
        final var increaseValue = decreaseIntegerMutation.getDecreaseBy().getValue();

        return String.format(decreaseIntTemplate, sanitisedVarName, increaseValue);
    }

    private void throwIfChapterVariable(final GameVariableLevel level) {
        if (level != GameVariableLevel.GAME) {
            throw new UnsupportedOperationException();
        }
    }
}
