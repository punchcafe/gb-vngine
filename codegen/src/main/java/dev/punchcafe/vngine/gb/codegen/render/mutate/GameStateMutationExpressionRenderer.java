package dev.punchcafe.vngine.gb.codegen.render.mutate;

import dev.punchcafe.vngine.gb.codegen.csan.CNameSanitiser;
import dev.punchcafe.vngine.pom.model.VariableTypes;
import dev.punchcafe.vngine.pom.model.vngml.*;

public class GameStateMutationExpressionRenderer implements GameStateMutationExpressionVisitor<String> {

    private static String GAME_STATE_POINTER_ARG = "game_state";

    @Override
    public String visitSetBooleanMutation(SetBooleanMutation setBooleanMutation) {
        final var methodName = MutationMethodNameConverter.convert(setBooleanMutation);

        final var targetVariable = setBooleanMutation.getVariableToMutate();
        // assume all variable are GLOBAL
        final var variableName = CNameSanitiser.sanitiseVariableName(targetVariable.getVariableName(),
                VariableTypes.BOOL,
                targetVariable.getGameVariableLevel());
        final var targetValue = setBooleanMutation.getChangeValue().isValue() ? 1 : 0;
        final var methodBody = String.format("%s->%s = %d;", GAME_STATE_POINTER_ARG, variableName, targetValue);

        return createPredicateMethodWithNameAndBody(methodName, methodBody);
    }

    @Override
    public String visitSetStringMutation(SetStringMutation setStringMutation) {
        final var methodName = MutationMethodNameConverter.convert(setStringMutation);

        final var targetVariable = setStringMutation.getVariableToMutate();
        // assume all variable are GLOBAL
        final var variableName = CNameSanitiser.sanitiseVariableName(targetVariable.getVariableName(),
                VariableTypes.STR,
                targetVariable.getGameVariableLevel());
        final var targetValue = setStringMutation.getChangeValue().getValue();
        final var methodBody = String.format("%s->%s = \"%s\";", GAME_STATE_POINTER_ARG, variableName, targetValue);

        return createPredicateMethodWithNameAndBody(methodName, methodBody);
    }

    @Override
    public String visitIncreaseIntegerMutation(IncreaseIntegerMutation increaseIntegerMutation) {
        final var methodName = MutationMethodNameConverter.convert(increaseIntegerMutation);

        final var targetVariable = increaseIntegerMutation.getVariableToModify();
        // assume all variable are GLOBAL
        final var variableName = CNameSanitiser.sanitiseVariableName(targetVariable.getVariableName(),
                VariableTypes.INT,
                targetVariable.getGameVariableLevel());
        final var targetValue = increaseIntegerMutation.getIncreaseBy().getValue();
        final var methodBody = String.format("%s->%s += %d;", GAME_STATE_POINTER_ARG, variableName, targetValue);

        return createPredicateMethodWithNameAndBody(methodName, methodBody);
    }

    @Override
    public String visitDecreaseIntegerMutation(DecreaseIntegerMutation decreaseIntegerMutation) {
        final var methodName = MutationMethodNameConverter.convert(decreaseIntegerMutation);

        final var targetVariable = decreaseIntegerMutation.getVariableToModify();
        // assume all variable are GLOBAL
        final var variableName = CNameSanitiser.sanitiseVariableName(targetVariable.getVariableName(),
                VariableTypes.INT,
                targetVariable.getGameVariableLevel());
        final var targetValue = decreaseIntegerMutation.getDecreaseBy().getValue();
        final var methodBody = String.format("%s->%s -= %d;", GAME_STATE_POINTER_ARG, variableName, targetValue);

        return createPredicateMethodWithNameAndBody(methodName, methodBody);
    }

    private String createPredicateMethodWithNameAndBody(final String name, final String body) {
        return String.format("void %s(struct GameState * " + GAME_STATE_POINTER_ARG + ")" +
                "\n{" +
                "\n    %s" +
                "\n}", name, body);
    }
}
