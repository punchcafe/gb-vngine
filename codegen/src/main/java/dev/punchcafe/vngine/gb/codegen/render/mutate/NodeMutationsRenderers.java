package dev.punchcafe.vngine.gb.codegen.render.mutate;

import dev.punchcafe.vngine.gb.codegen.csan.NodeIdSanitiser;
import dev.punchcafe.vngine.gb.codegen.csan.NodeMutationsArrayName;
import dev.punchcafe.vngine.gb.codegen.render.ComponentRenderer;
import dev.punchcafe.vngine.pom.model.Node;
import dev.punchcafe.vngine.pom.model.ProjectObjectModel;
import dev.punchcafe.vngine.pom.parse.vngml.GameStateMutationExpressionParser;
import lombok.Builder;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static dev.punchcafe.vngine.gb.codegen.render.ComponentRendererName.NODE_MUTATION_RENDERER_NAME;
import static dev.punchcafe.vngine.gb.codegen.render.mutate.GameStateMutationRenderer.GAME_STATE_MUTATION_RENDERER;
import static java.util.stream.Collectors.joining;

@Builder
public class NodeMutationsRenderers implements ComponentRenderer {

    private static final GameStateMutationExpressionParser mutationParser = new GameStateMutationExpressionParser();

    private final ProjectObjectModel<?> gameConfig;

    @Override
    public String render() {
        return gameConfig.getChapterConfigs().get(0)
                .getNodes()
                .stream()
                .map(this::renderNodeMutations)
                .collect(Collectors.joining("\n"));
    }

    @Override
    public List<String> dependencies() {
        return List.of(GAME_STATE_MUTATION_RENDERER);
    }

    @Override
    public String componentName() {
        return NODE_MUTATION_RENDERER_NAME;
    }

    private String renderNodeMutations(final Node node) {

        if(node.getGameStateModifiers() == null || node.getGameStateModifiers().isEmpty()){
            return "";
        }

        final var mutationArrayPointerName = NodeMutationsArrayName.getMutationArrayForNode(node);
        final var methodReferencesArray = node.getGameStateModifiers().stream()
                .map(mutationParser::parseExpression)
                .map(MutationMethodNameConverter::convert)
                .map(methodName -> "&" + methodName)
                .collect(joining(",", "{", "}"));
        final var numberOfMutations = node.getGameStateModifiers().size();

        return String.format("void (*%s[%d])(struct GameState *) = %s;", mutationArrayPointerName,
                numberOfMutations,
                methodReferencesArray);
    }

}
