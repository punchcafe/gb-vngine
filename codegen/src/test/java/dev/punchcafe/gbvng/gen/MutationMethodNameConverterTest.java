package dev.punchcafe.gbvng.gen;


import dev.punchcafe.gbvng.gen.render.mutate.MutationMethodNameConverter;
import dev.punchcafe.vngine.pom.model.vngml.DecreaseIntegerMutation;
import dev.punchcafe.vngine.pom.model.vngml.IncreaseIntegerMutation;
import dev.punchcafe.vngine.pom.model.vngml.SetBooleanMutation;
import dev.punchcafe.vngine.pom.model.vngml.SetStringMutation;
import dev.punchcafe.vngine.pom.model.vngpl.variable.GameVariableLevel;
import dev.punchcafe.vngine.pom.model.vngpl.variable.bool.BoolGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.bool.BooleanLiteral;
import dev.punchcafe.vngine.pom.model.vngpl.variable.integer.IntegerGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.integer.IntegerLiteral;
import dev.punchcafe.vngine.pom.model.vngpl.variable.string.StringGameVariable;
import dev.punchcafe.vngine.pom.model.vngpl.variable.string.StringLiteral;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import static io.github.jsonSnapshot.SnapshotMatcher.expect;
import static io.github.jsonSnapshot.SnapshotMatcher.start;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class MutationMethodNameConverterTest {

    @BeforeAll
    void beforeAll() {
        start(Snapshot::asJsonString);
    }

    @Test
    void canHashIncreaseIntegerMutation(){
        final var mutation = IncreaseIntegerMutation.builder()
                .variableToModify(new IntegerGameVariable(GameVariableLevel.GAME, "my-variable"))
                .increaseBy(new IntegerLiteral(50))
                .build();
        expect(MutationMethodNameConverter.convert(mutation)).toMatchSnapshot();
    }

    @Test
    void canHashDecreaseIntegerMutation(){
        final var mutation = DecreaseIntegerMutation.builder()
                .variableToModify(new IntegerGameVariable(GameVariableLevel.GAME, "my-variable"))
                .decreaseBy(new IntegerLiteral(101))
                .build();
        expect(MutationMethodNameConverter.convert(mutation)).toMatchSnapshot();
    }

    @Test
    void canHashSetBooleanMutationToTrue(){
        final var mutation = SetBooleanMutation.builder()
                .variableToMutate(new BoolGameVariable(GameVariableLevel.GAME, "my-bool"))
                .changeValue(BooleanLiteral.TRUE)
                .build();
        expect(MutationMethodNameConverter.convert(mutation)).toMatchSnapshot();
    }

    @Test
    void canHashSetBooleanMutationToFalse(){
        final var mutation = SetBooleanMutation.builder()
                .variableToMutate(new BoolGameVariable(GameVariableLevel.GAME, "mybol"))
                .changeValue(BooleanLiteral.FALSE)
                .build();
        expect(MutationMethodNameConverter.convert(mutation)).toMatchSnapshot();
    }

    @Test
    void canHashSetStringMutation(){
        final var mutation = SetStringMutation.builder()
                .variableToMutate(new StringGameVariable(GameVariableLevel.GAME, "my-string"))
                .changeValue(new StringLiteral("newval"))
                .build();
        expect(MutationMethodNameConverter.convert(mutation)).toMatchSnapshot();
    }

    @Test
    void canHashSetStringMutationForComplextString(){
        final var mutation = SetStringMutation.builder()
                .variableToMutate(new StringGameVariable(GameVariableLevel.GAME, "my-string"))
                .changeValue(new StringLiteral("Hello! I'm using lot's of really @nnoyig pun(tu4t`|0|\\|"))
                .build();
        expect(MutationMethodNameConverter.convert(mutation)).toMatchSnapshot();
    }
}
