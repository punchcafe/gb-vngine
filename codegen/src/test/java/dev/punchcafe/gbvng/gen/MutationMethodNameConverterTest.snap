MutationMethodNameConverterTest.canHashDecreaseIntegerMutation=[
  "dec_game_var_int_my_variable_to_101"
]


MutationMethodNameConverterTest.canHashIncreaseIntegerMutation=[
  "inc_game_var_int_my_variable_to_50"
]


MutationMethodNameConverterTest.canHashSetBooleanMutationToFalse=[
  "set_game_var_bool_mybol_to_false"
]


MutationMethodNameConverterTest.canHashSetBooleanMutationToTrue=[
  "set_game_var_bool_my_bool_to_true"
]


MutationMethodNameConverterTest.canHashSetStringMutation=[
  "set_game_var_str_my_string_to_newval"
]


MutationMethodNameConverterTest.canHashSetStringMutationForComplextString=[
  "set_game_var_str_my_string_to_Hello__URLCHAR_21____PLUS__I__URLCHAR_27__m__PLUS__using__PLUS__lot__URLCHAR_27__s__PLUS__of__PLUS__really__PLUS____URLCHAR_40__nnoyig__PLUS__pun__URLCHAR_28__tu4t__URLCHAR_60____PERCENT__7C0__PERCENT__7C__PERCENT__5C__PERCENT__7C"
]


dev.punchcafe.gbvng.gen.MutationMethodNameConverterTest.canHashDecreaseIntegerMutation=[
  "dec_game_var_int_my_variable_to_101"
]


dev.punchcafe.gbvng.gen.MutationMethodNameConverterTest.canHashIncreaseIntegerMutation=[
  "inc_game_var_int_my_variable_to_50"
]


dev.punchcafe.gbvng.gen.MutationMethodNameConverterTest.canHashSetBooleanMutationToFalse=[
  "set_game_var_bool_mybol_to_false"
]


dev.punchcafe.gbvng.gen.MutationMethodNameConverterTest.canHashSetBooleanMutationToTrue=[
  "set_game_var_bool_my_bool_to_true"
]


dev.punchcafe.gbvng.gen.MutationMethodNameConverterTest.canHashSetStringMutation=[
  "set_game_var_str_my_string_to_newval"
]


dev.punchcafe.gbvng.gen.MutationMethodNameConverterTest.canHashSetStringMutationForComplextString=[
  "set_game_var_str_my_string_to_Hello__URLCHAR_21____PLUS__I__URLCHAR_27__m__PLUS__using__PLUS__lot__URLCHAR_27__s__PLUS__of__PLUS__really__PLUS____URLCHAR_40__nnoyig__PLUS__pun__URLCHAR_28__tu4t__URLCHAR_60____PERCENT__7C0__PERCENT__7C__PERCENT__5C__PERCENT__7C"
]